src/Tmds.Ssh/KerberosCredential.cs

```diff
@@ -0,0 +1,339 @@
+            // Windows client support. Sign only will pass in SECQOP_WRAP_NO_ENCRYPT
+            // to MakeSignature which fails.
+            // https://github.com/dotnet/runtime/issues/103461
+            RequiredProtectionLevel = ProtectionLevel.EncryptAndSign,
+            // While RFC states this should be set to "false", Win32-OpenSSH
+            // fails if it's not true. I'm unsure if openssh-portable on Linux
+            // will fail in the same way or not.
+            RequireMutualAuthentication = true,
+            TargetName = spn,
+        };
+
+        using var authContext = new NegotiateAuthentication(negotiateOptions);
+        bool isAuthSuccess = await TryStageAuthentication(connection, logger, authContext, ct).ConfigureAwait(false);
+        if (!isAuthSuccess)
+        {
+            return false;
+        }
+
+        try
+        {
+            using var message = authContext.IsSigned
+                ? CreateGssapiMicData(connection.SequencePool, connectionInfo.SessionId!,  userName, authContext)
+                : CreateGssapiCompleteMessage(connection.SequencePool);
+            await connection.SendPacketAsync(message.Move(), ct).ConfigureAwait(false);
+        }
+        catch (MissingMethodException)
+        {
+            // Remove once .NET 8 is no longer the minimum and we get rid of reflection.
+            return false;
+        }
+
+        return true;
+    }
+
+    private static async Task<bool> TryStageOid(SshConnection connection, ILogger logger, string userName, CancellationToken ct)
+    {
+        {
+            using var userAuthMsg = CreateOidRequestMessage(connection.SequencePool,
+                                        userName, KRB5_OID);
+            await connection.SendPacketAsync(userAuthMsg.Move(), ct).ConfigureAwait(false);
+        }
+
+        while (true)
+        {
+            using Packet response = await connection.ReceivePacketAsync(ct).ConfigureAwait(false);
+            MessageId? mid = response.MessageId;
+
+            if (mid == MessageId.SSH_MSG_USERAUTH_BANNER)
+            {
+                // First message in exchange could be a banner which we ignore
+                // for now.
+                continue;
+            }
+
+            if (mid == MessageId.SSH_MSG_USERAUTH_FAILURE)
+            {
+                logger.AuthenticationKerberosFailed("No OID response received");
+                return false;
+            }
+            else if (mid != MessageId.SSH_MSG_USERAUTH_GSSAPI_RESPONSE)
+            {
+                ThrowHelper.ThrowProtocolUnexpectedValue();
+            }
+
+            ReadOnlySequence<byte> oidResponse = GetGssapiOidResponse(response);
+            if (KRB5_OID.AsSpan().SequenceEqual(oidResponse.IsSingleSegment ? oidResponse.FirstSpan : oidResponse.ToArray()))
+            {
+                return true;
+            }
+            else
+            {
+                var reader = new AsnReader(oidResponse.ToArray(), AsnEncodingRules.DER);
```

Comment:
If KRB5_OID is already DER encoder, why do we need this branch?

Reply:
It's to log the response OID from the server when it doesn't match what we sent. I'm not sure if this could happen but it's to verify the server is responding the way we would expect and to provide a way to see what happens in the failure scenarios through the logs.

Reply:
Ah yes, this is for logging.

I'd rather avoid the `AsnReader` here (to avoid unexpected issues on an unlikely branch), can you change it to `Convert.ToBase64String` the `oidResponse`?

Reply:
Sure thing.